### PR TITLE
Respect canFuseOn{CPU,GPU} in TE fuser

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43967 Respect canFuseOn{CPU,GPU} in TE fuser**

Differential Revision: [D23469048](https://our.internmc.facebook.com/intern/diff/D23469048)

### CODE REVIEW

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 1]
```diff
@@ -466,6 +467,34 @@ class TensorExprFuser {
     return true;
   }
 
+  bool canFuseOnDevice(Value* v) {
+    auto type = v->type()->cast<TensorType>();
+    if (!type) {
+      return true;
+    }
+    auto device = type->device();
+    if (!device) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:476]: I know it's an existing logic, but I think we might want to be stricter in how we treat tensors with no device in the type, at least in the NNC fuser.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 2]
```diff
@@ -530,10 +559,20 @@ class TensorExprFuser {
       REQ(producer->input(0)->node()->kind() == prim::ListConstruct);
       REQ(producer->input(0)->uses().size() == 1);
       REQ(producer->input(1)->node()->kind() == prim::Constant);
+      auto const& listConstruct = producer->input(0)->node();
+      for (auto const& input : listConstruct->inputs()) {
+        REQ(isFusableOnDevice(input->node()));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:564]: Not necessarily to do in this PR, but we also need to be checking that we're pulling in values/nodes with the same device.

**Review by ZolotukhinM — APPROVED** [review]
Looks good!

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 2 above)
Comment (bertmaher) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:564]: I'll have to think a bit more about restricting merges to same device.  I think that problem exists even without this PR, so I'm going to go ahead and land this one while I work on same-device.

**Pull request closed without merging**